target/linux/lantiq/files/arch/mips/lantiq/b43-sprom.c

```diff
@@ -0,0 +1,1196 @@
+		if (bus->host_pci->bus->number != priv->pci_bus ||
+		    PCI_SLOT(bus->host_pci->devfn) != priv->pci_dev)
+			pr_warn("ssb_fallback_sprom: pci bus/device num "
+				"mismatch: expected %i/%i, but got %i/%i\n",
+				priv->pci_bus, priv->pci_dev,
+				bus->host_pci->bus->number,
+				PCI_SLOT(bus->host_pci->devfn));
+		if (priv->devid_override)
+			bus->host_pci->device = priv->sprom.dev_id;
+		memcpy(out, &priv->sprom, sizeof(struct ssb_sprom));
+		return 0;
+	} else {
+		pr_err("%s: unable to fill SPROM for given bustype.\n",
+		       __func__);
+		return -EINVAL;
+	}
+}
+#endif /* CONFIG_SSB_PCIHOST */
+
+#if defined(CONFIG_BCMA_HOST_PCI)
+int b43_get_fallback_bcma_sprom(struct bcma_bus *bus, struct ssb_sprom *out)
+{
+	struct b43_sprom *priv = &b43_sprom;
+
+	if (bus->hosttype == BCMA_HOSTTYPE_PCI) {
+		if (bus->host_pci->bus->number != priv->pci_bus ||
+		    PCI_SLOT(bus->host_pci->devfn) != priv->pci_dev)
+			pr_warn("bcma_fallback_sprom: pci bus/device num "
+				"mismatch: expected %i/%i, but got %i/%i\n",
+				priv->pci_bus, priv->pci_dev,
+				bus->host_pci->bus->number,
+				PCI_SLOT(bus->host_pci->devfn));
+		if (priv->devid_override)
+			bus->host_pci->device = priv->sprom.dev_id;
+		memcpy(out, &priv->sprom, sizeof(struct ssb_sprom));
+		return 0;
+	} else {
+		pr_err("%s: unable to fill SPROM for given bustype.\n",
+		       __func__);
+		return -EINVAL;
+	}
+}
+#endif /* CONFIG_BCMA_HOST_PCI */
+
+static int b43_sprom_probe(struct platform_device *pdev)
+{
+	
+	struct device *dev = &pdev->dev;
+	struct device_node *node = dev->of_node;
+	struct b43_sprom *priv = &b43_sprom;
+	int ret;
+	pr_warn("b43 sprom probe is launched");
+	priv->dev = dev;
+
+	b43_sprom_set(priv);
+
+	of_property_read_u32(node, "pci-bus", &priv->pci_bus);
+	of_property_read_u32(node, "pci-dev", &priv->pci_dev);
+
+	of_get_mac_address(node, priv->mac);
+	if (is_valid_ether_addr(priv->mac)) {
+		dev_info(dev, "mtd mac %pM\n", priv->mac);
+	} else {
+		random_ether_addr(priv->mac);
+		dev_info(dev, "random mac %pM\n", priv->mac);
+	}
+	
+	memcpy(priv->sprom.il0mac, priv->mac, ETH_ALEN);
+ 	memcpy(priv->sprom.et0mac, priv->mac, ETH_ALEN);
+ 	memcpy(priv->sprom.et1mac, priv->mac, ETH_ALEN);
+
+#if defined(CONFIG_SSB_PCIHOST)
+	ret = ssb_arch_register_fallback_sprom(&b43_get_fallback_ssb_sprom);
+	if (ret)
+		return ret;
+#endif /* CONFIG_SSB_PCIHOST */
+
+#if defined(CONFIG_BCMA_HOST_PCI)
+	ret = bcma_arch_register_fallback_sprom(b43_get_fallback_bcma_sprom);
+	if (ret)
+		return ret;
+#endif /* CONFIG_BCMA_HOST_PCI */
+
+	return 0;
+}
+
+static const struct of_device_id b43_sprom_of_match[] = {
+	{ .compatible = "brcm,bcm43xx-sprom", .data = NULL },
+#if defined (CONFIG_SSB_PCIHOST)
+	{ .compatible = "brcm,bcm4306-sprom", .data = &bcm4306_sprom, },
+	{ .compatible = "brcm,bcm4321-sprom", .data = &bcm4321_sprom, },
+	{ .compatible = "brcm,bcm4322-sprom", .data = &bcm4322_sprom, },
+	{ .compatible = "brcm,bcm43222-sprom", .data = &bcm43222_sprom, },
```

Comment:
How does the vendor FW read the nvram?

Reply:
It's a good question. As far as I could see, the device does not have a physical SPROM so my guess is that it may be located somewhere in the flash. Unfortunetly, obfuscation doesn't make this job easy and I couldn't fin exacly where, but the default map provided in the SPROM fallback driver does the job.